src/devices/cddasongloader.cpp

```diff
@@ -214,13 +304,9 @@ void CddaSongLoader::AudioCDTagsLoaded(
 }
 
 bool CddaSongLoader::HasChanged() {
-  if ((cdio_ && cdda_) && cdio_get_media_changed(cdio_) != 1) {
-    return false;
-  }
-  // Check if mutex is already token (i.e. init is already taking place)
-  if (!mutex_load_.tryLock()) {
```

Comment:
From cddadevice.cpp:
```
void CddaDevice::Init() {
  song_count_ = 0;  // Reset song count, in case it was already set
  cdda_song_loader_.LoadSongs();
}

void CddaDevice::Refresh() {
  if (!cdda_song_loader_.HasChanged()) {
    return;
  }
  Init();
}
```
So, the mutex is preventing the LoadSongs from being when there's a loading thread running. That synchronization logic really shouldn't be spanning two classes, but it looks like it was the result of CddaDevice being split into two pieces. But if you do want to stick with this existing pattern, then I think you could check the state of loading_future_ here.

The old code also has a timing hole. You could potentially call Refresh multiple times in succession and kick off several threads before the first thread has a chance to grab the lock. I think that could also be fixed by checking the future object.

Reply:
I think this could neatly be adressed by my idea of making `CddaSongLoader` more stateful and encapsulating the logic for only re-reading data if the medium was changed. `CddaDevice::Refresh` could then just call `Init` without any checks there

Reply:
I have now encapsulated checks for `HasChanged` in  `CddaSongLoader::LoadSongs`. Subsequent calls to `LoadSongs` without a media change will not perform any reading but simply emit a signal with the complete track list with all metadata once. `CddaDevice::Refresh` therefore can just call `::Init()` now and no more synchronisation issues can occur there.